### ISSUE TITLE
Problems with ProblemDetails

### ISSUE THREAD

**Issue by lurumad** [Microsoft/aspnet-api-versioning#318]
Hi,

I have an annoying bug in my API when I use ApiVersioning and ProblemDetails, and I do not know which is the correct repo (this or aspnet team) where I have to create this issue. 

If I register ProblemDetails before ApiVersioning, the InvalidModelStateResponseFactory options does not work:

```csharp
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .Configure<ApiBehaviorOptions>(options =>
            {
                options.InvalidModelStateResponseFactory = context =>
                {
                    var problemDetails = new ValidationProblemDetails(context.ModelState)
                    {
                        Instance = context.HttpContext.Request.Path,
                        Status = StatusCodes.Status400BadRequest,
                        Type = $"https://httpstatuses.com/400",
                        Detail = "Errors"
                    };
                    return new BadRequestObjectResult(problemDetails)
                    {
                        ContentTypes =
                        {
                            "application/problem+json",
                            "application/problem+xml"
                        }
                    };
                };
            })
            .AddApiVersioning(
                options =>
                {
                    options.AssumeDefaultVersionWhenUnspecified = true;
                    options.ReportApiVersions = true;
                    options.DefaultApiVersion = new ApiVersion(1, 0);
                }
            )
            .AddMvcCore()
            .AddJsonFormatters()
            .AddApiExplorer()
            .AddVersionedApiExplorer()
            .AddAuthorization()
            .AddDataAnnotations();
    }

    public void Configure(IApplicationBuilder app, IHostingEnvironment env)
    {
        app.UseMvc();
    }
}
```
```json
{
    "": [
        "The input was not valid."
    ]
}
```
```text
content-type →application/json; charset=utf-8
date →Fri, 20 Jul 2018 11:42:39 GMT
server →Kestrel
transfer-encoding →chunked
x-powered-by →ASP.NET
x-sourcefiles →=?UTF-8?B?YzpcdXNlcnNcbHJ1aXpcc291cmNlXHJlcG9zXFdlYkFwcGxpY2F0aW9uNlxXZWJBcHBsaWNhdGlvbjZcYXBpXHRlc3Q=?=
```

if I change the order:

```json
{
    "errors": {
        "": [
            "The input was not valid."
        ]
    },
    "type": "https://httpstatuses.com/400",
    "title": "One or more validation errors occurred.",
    "status": 400,
    "detail": "Errors",
    "instance": "/api/test"
}
```
```text
content-type →application/problem+json; charset=utf-8
date →Fri, 20 Jul 2018 11:44:23 GMT
server →Kestrel
transfer-encoding →chunked
x-powered-by →ASP.NET
x-sourcefiles →=?UTF-8?B?YzpcdXNlcnNcbHJ1aXpcc291cmNlXHJlcG9zXFdlYkFwcGxpY2F0aW9uNlxXZWJBcHBsaWNhdGlvbjZcYXBpXHRlc3Q=?=
```

I don't know why the order of the services matters.

Regards!

**Comment by commonsensesoftware** [comment]
The issue is likely here. In general, I agree - the order of service registration should not matter. Despite what one might think, there are some interesting albeit edge cases where registration order matters.

Although it's never been reported before, I recently noticed that the service registration was less than ideal in the API versioning libraries. Some things were not as _lazy_ as they should be or unnecessarily re-registered potential dependencies (ex: `services.AddMvcCore()`).  These will be all addressed in the forthcoming v3.0. I'd show you the change now, but it's still part of an unstable working set that I haven't pushed yet. It should appear in the `v3` branch in the next few days.

As a general rule, I would suggest registering things in a logical dependency order, even though it shouldn't necessarily be a requirement. For example, `AddMvcCore()` or `AddMvc()` should be one of the first things registered.

**Comment by commonsensesoftware** [comment]
The v3.0 beta has been published. If you wouldn't mind trying it out on your configuration to see if it resolves this issue, I'm sure it would be useful to all. The DI setups are deferred as long as possible now, so service registration order _should_ not matter. If your issue is resolved, this can be closed out; otherwise, I'll evaluate what else can be done to improve the setup.  Thanks.

**Comment by commonsensesoftware** [comment]
Were you able to test out the beta to determine if the issue was resolved? I suspect it will be.

**Comment by commonsensesoftware** [comment]
Prior to 3.0, the registration order _may_ be a factor. This should definitely be solved in 3.0+. The workaround for earlier versions will be to register things in order. I seems like you've come to the same conclusion by updating to the latest 3.0 prerelease or your registration order. Feel free to reopen this issue if you didn't not come up with a solution. Thanks.
